### DISCUSSION TITLE
Is it recommended to automatically call 'build' before 'dev' during development?

### DISCUSSION THREAD

**Discussion by nicognaW** [remix-run/remix#6142] [Q&A] [answered]
I find it quite tiring to manually call build before dev every time during development. I am considering setting up a process to automatically call build before dev to streamline the workflow. I'm posting because I'm unsure if there are any potential side effects to this approach.

I asked GPT-4, and it says:

> In the Remix framework, running npm run build before npm run dev is necessary because the build process generates client-side assets and optimized bundles, which are used by the development server. Unlike frameworks like Next.js, Remix has a separate build step to provide more control and optimization.
>
> Automatically calling "build" before "dev" increases startup time, can result in unnecessary builds, and consumes more system resources. To mitigate these side effects, you can use watch mode (if supported) to rebuild only affected assets, implement conditional checks to determine if a build is necessary, or run the build process manually when needed.
>
> You can modify your package.json scripts to run the build before dev automatically. However, consider the potential side effects and choose the approach that best suits your workflow, balancing the need for an updated build with a fast, efficient development experience.

I wonder if performance is the only sacrifice if I call `build` every time. I would consider using Turborepo to circumvent it.

**Comment by sergiodxa** [comment, marked as answer]
You can do it to ensure there's always a build folder, but it's not really necessary. If you want to do it, it can be done as simple as `npm run build && npm run dev`.

**Reply by kiliman** [reply]
If I'm not mistaken, the reason to run `npm run build` first was so that if you're using Tailwind, it would generate the *tailwind.css* file before Remix would try to import it in your *root.tsx*. You technically only had to do this one time.

Now that Remix supports Tailwind directly, you don't have to worry about this pre-build step. Just follow the instructions in the Tailwind Guide.

https://remix.run/docs/en/1.15.0/guides/styling#tailwind-css

**Reply by MichaelDeBoey** [reply]
@kiliman It is still necessary if you're building your `server.ts` file, like we do in `blues-stack` (which I took over in #6145)

**Reply by kiliman** [reply]
I do something similar in my project. But just like yours, you're calling `build:server` inside your `dev:*` scripts. I think the OP is asking why it's recommended to call `npm run build` followed by `npm run dev`.

You should only have to call `npm run dev` (since it will build the initial server.js file).

**Reply by MichaelDeBoey** [reply]
`dev:server` indeed runs `build:server`, but adds an extra `--watch` flag.
Since all `dev:*` run in parallel, it sometimes happens `dev:server`/`build:server` hasn't run yet before it's needed, so that's why I do an extra `build:server` before running `dev:*`
